### ISSUE TITLE
Check/064: Test reports inaccurate caret/ligature count

### ISSUE THREAD

**Issue by eliheuer** [googlefonts/fontbakery#1976]
This issue was discussed in another issue thread, but I'm moving it to it's own issues so it doesn't get lost there: https://github.com/googlefonts/fontbakery/issues/1923#issuecomment-404595024

## Observed behaviour

Testing Dosis ( https://github.com/eliheuer/Dosis/commit/fb1e2a004fcb93e5437653dae09a7c1cc8e8dda6 ) with `fontbakery` I get a `Check/064 WARN`, but the information is inaccurate. This file has seven ligatures, all with carets, some having two carets per ligature, `f_f_i` for example.  

## Expected behaviour

Accurate caret/ligature count reported.

## Resources and exact process needed to replicate

Please see the comment dave made here: https://github.com/googlefonts/fontbakery/issues/1923#issuecomment-404595024

He suggests this needs work to fix the issue:
https://github.com/googlefonts/fontbakery/blob/0f82289da2f0e384719b498435b6a612115d1992/Lib/fontbakery/specifications/gdef.py#L42-L59

**Comment by felipesanches** [comment]
thanks, @eliheuer !

**Comment by mjlagattuta** [comment]
I have been experiencing this as well when testing Roboto Slab. This check appears be checking glyphs with an underscore in the name (i.e. brevecomb_acutecomb) but not things like fi, fl, or f_f_i and f_f_l when they are named based on their unicode value. 

Interestingly enough, `Check/065` detects fi and fl as well as other non-ligature glyphs like stacked accents, fractions, and the Catalan "Ldot" but misses things like f_f_i and f_f_l

GPOS table lacks kerning info for the following non-ligated sequences: ['uni030A_acutecomb', 'uni0302_acutecomb', 'uni0302_gravecomb', 'uni0302_hookabovecomb', 'uni0302_tildecomb', 'uni0306_acutecomb', 'uni0306_gravecomb', 'uni0306_hookabovecomb', 'uni0306_tildecomb', 'l_periodcentered', 'L_periodcentered', 'one_slash', 'three_slash', 'five_slash', 'seven_slash', 'N_o', 'f_i', 'f_l'] [code: lacks-kern-info]

**Comment by felipesanches** [comment]
This seems to be the problem:
https://github.com/googlefonts/fontbakery/blob/0f82289da2f0e384719b498435b6a612115d1992/Lib/fontbakery/specifications/gdef.py#L49

The comparison in the line above seems to assume a single caret-position for each ligature. But I think we can have multiple caret positing for a single ligature glyph such as f_f_i.

**Comment by felipesanches** [comment]
I am considering a different implementation that makes sure at least one caret position is defined for each ligature glyph, instead of simply relying on the total count, which can be misleading.

**Comment by eliheuer** [comment]
That sounds good. But for a more accurate test, could we just check to make sure the number of caret positions is the same as the number of underscores?

**Comment by eliheuer** [comment]
Ok, I guess the only other way to do this would be to compile a list of expected ligature counts.

Are we ok with ligatures sometimes having improper caret positions?

One example I have found is `allah-ar` in Markazi Text. It currently has improperly set carets but would pass a test that just checks for one.

![caret](https://user-images.githubusercontent.com/5162664/44593864-97e17880-a792-11e8-9a9b-1735d52faa27.png)

الله is made up of ا ل ل ه (or Alif Lam Lam Ha, from right to left), try pasting الله into the text preview for [Amiri](https://fonts.google.com/?subset=arabic&query=amiri), and then try to separate the characters with spaces. You will see the carets function as expected. But do the same this for [Markazi](https://fonts.google.com/?subset=arabic&query=Markazi) and note the the caret functionality is broken.

**Comment by felipesanches** [comment]
This one was tricky! But now I got it sorted out. This is the new output. It is not only fixed (because the previous info was non-sense) but it is also more informative actually listing the glyph names instead of simply relying on a count of glyphs and of caret positioning data.

![screenshot at 2018-08-24 18 43 48](https://user-images.githubusercontent.com/213676/44609424-136d1500-a7e7-11e8-9e84-b45407659913.png)

**Comment by felipesanches** [comment]
... but I need feedback on this, since it seems that one would not want to have caret positioning info for diacritics.

**Comment by davelab6** [comment]
> Ok, I guess the only other way to do this would be to compile a list of expected ligature counts.

 we already have a list of expected contours per glyph

**Comment by felipesanches** [comment]
The important thig to address here is what @moyogo said at https://github.com/googlefonts/fontbakery/issues/1923#issuecomment-398891738:

> Most importantly, ligature caret don’t make sense in all ligatures. It makes little sense to have ligature carets for ligature formed of base glyph + non spacing diacritic. You don’t want ligature carets for Indic clusters either, like क्ष as a ka_virama_ssa ligature for example.

How can we automate this to algorithmically distinguish base glyphs from non spacing diacritics?!

**Comment by davelab6** [comment]
> How can we automate this to algorithmically distinguish base glyphs from non spacing diacritics?!

I suggest reading the opentype specifications, if you have not already done so. https://docs.microsoft.com/en-us/typography/opentype/spec/

**Comment by felipesanches** [comment]
I surely did! But it is always good to go back to it form time to time :-D

**Comment by eliheuer** [comment]
Why don't we just make a very basic list of expected caret counts and implement a check for only what is on the list? That way we could get the list started with minimal effort and then slowly improve it as bugs come in?

**Comment by mjlagattuta** [comment]
I agree that a list would be good for rejecting combining diacritics etc.

For getting the correct number of caret positioning values, could the check not run through `LigatureSubst` tags and for each child `Ligature` tag set the expected ligature caret value equal to the number of `components`?
```
<LigatureSubst index="0" Format="1">
  <LigatureSet glyph="f">
    <Ligature components="f,i" glyph="f_f_i"/>
    <Ligature components="f,l" glyph="f_f_l"/>
    <Ligature components="f" glyph="f_f"/>
    <Ligature components="i" glyph="fi"/>
    <Ligature components="l" glyph="fl"/>
  </LigatureSet>
</LigatureSubst>
```

Since the number of component glyphs is always one less than the total number of glyphs, it should always be the equal to the number of expected caret values.

**Comment by eliheuer** [comment]
Counting the ligature components is a great idea! I made a pull request to gftools with the start of a tool that will do this: 

https://github.com/googlefonts/gftools/pull/79

**Comment by felipesanches** [comment]
OK. with the commits above, we don't get diacritics anymore, because I believe I am now correctly looking up the Font Feature index for kern data and for ligatures and then using such index values to only get data from the relevant Lookups.

**Comment by felipesanches** [comment]
I moved the improvement proposal into a new issue as I consider this one (Test reports inaccurate caret/ligature count) to be now properly fixed.

**Comment by khaledhosny** [comment]
> الله is made up of ا ل ل ه (or Alif Lam Lam Ha, from right to left), try pasting الله into the text preview for Amiri, and then try to separate the characters with spaces. You will see the carets function as expected. But do the same this for Markazi and note the the caret functionality is broken.

This is unlikely to have anything to do with the GDEF caret positions. It works with both fonts in Firefox, but I don’t know which browser you tested with. Note that Amiri does not have a ligature here at ll, but 4 separate glyphs.
